docs/design/2021-04-18-common-table-expression.md

```diff
@@ -0,0 +1,287 @@
+Also DataSource will be converted to `PhysicalTempTableReader`.
+
+The Physical Plan will be like:
+
+![physical_plan](./imgs/physical_plan.png)
+
+#### Build Executor
+Three structures will be constructed:
+1. `CTEExec`: Evaluate seed part and recursive part iteratively.
+2. `TempTableReaderExec`: Read result of previous iteration and return result to parent operator.
+3. `CTEStorage`: This is where the materialized results stored. `CTEExec` will write it and `TempTableReaderExec` will read it.
+
+The executor tree will be like:
+
+![executors](./imgs/executors.png)
+
+#### Execution
+The implementation of `CTEExec` is as follows. It first checks whether `CTEStorage` is filled, if not, it will compute seed part and recursive part to fill `CTEStorage`.
+So the `CTEStorage` will only be filled by first executed `CTEExec`, and other `CTEExec` which references to the same `CTEStorage` just reads it.
+
+```Go
+func (e *CTEExec) Next(req *Chunk) {
+    // 1. the first executed CTEExec will be responsible to fill storage.
+    if !e.storage.Done() {
+        // 1.1 compute seed part.
+        // 1.2 compute recursive part iteratively.
+        e.storage.SetDone()
+    }
+    // 2. return chunk in e.storage
+}
+```
+
+The filling of `CTEStorage` is done by `CTEExec` and `TempTableReaderExec` together. The following figure describes the process.
+
+![cte_computation](./imgs/cte_computation.png)
+
+The data of iter_0 is generated by computing seed part and stored into `CTEStorage`. Data of iter_i is generated as follows:
+1. `CTEExec` call child's next() method, then `TempTableReaderExec` will read iter_i-1 data from `CTEStorage`.
+2. Data is returnd to `Selection`.
+3. `Selection` filter data and return it to `CTEExec`.
+4. `CTEExec` stores the result to `CTEStorage`, and the result labeled as iter_1 data.
+
+The above steps will continue to iterate until:
+1. No data is generated by recursive part.
+2. Iteration number reachs `cte_max_recursion_depth`.
+3. Execution time reachs `max_execution_time`
+
+## Test Design
+
+### Functional Tests
+
+1. Basic usage of non-recursive and recursive CTE.
+2. Define a new CTE within a CTE.
+3. Use CTE in subquery.
+4. Use CTE in UPDATE/DELETE/INSERT statements.
+5. CTE name conflicts with other table name.
+6. Join CTE with other tables/CTE.
+7. Use expression with CTE.
+
+### Scenario Tests
+
+We should test CTE used together with other features:
+1. Use CTE with PREPARE/EXECUTE/PlanCache.
+2. Use CTE with partition table.
+3. Stale read.
+4. Clustered index.
+
+### Compatibility Tests
+
+None
+
+### Benchmark Tests
+
+We should know the performance of recursive/non-recursive CTE compared with Oracle and MySQL.
+
+## Impacts & Risks
+
+CTE is a new feature and it will not affect the overall performance.
+But for now we only use materialization to implement non-recursive CTE. Performance may not be as good as MySQL in some scenarios.
```

Comment:
Delete `Performance may not be as good as MySQL in some scenarios`

Reply:
done